components/brave_wallet/browser/zcash/zcash_orchard_storage.cc

```diff
@@ -0,0 +1,438 @@
+                           " "
+                           "("
+                           "account_id TEXT NOT NULL PRIMARY KEY,"
+                           "account_birthday INTEGER NOT NULL,"
+                           "latest_scanned_block INTEGER NOT NULL,"
+                           "latest_scanned_block_hash TEXT NOT NULL);") &&
+         transaction.Commit();
+}
+
+bool ZCashOrchardStorage::UpdateSchema() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  return true;
+}
+
+std::optional<ZCashOrchardStorage::Error> ZCashOrchardStorage::RegisterAccount(
+    mojom::AccountIdPtr account_id,
+    uint64_t account_birthday_block,
+    const std::string& account_birthday_block_hash) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!EnsureDbInit()) {
+    NOTREACHED_IN_MIGRATION();
+    return Error{kDbInitError, "Failed to init database "};
+  }
+
+  sql::Transaction transaction(&database_);
+  if (!transaction.Begin()) {
+    NOTREACHED_IN_MIGRATION();
+    return Error{kDbInitError, "Failed to init database "};
+  }
+
+  sql::Statement register_account_statement(database_.GetCachedStatement(
+      SQL_FROM_HERE, "INSERT OR REPLACE INTO " kAccountMeta " "
+                     "(account_id, account_birthday, latest_scanned_block, "
+                     "latest_scanned_block_hash) "
+                     "VALUES (?, ?, ?, ?)"));
+
+  register_account_statement.BindString(0, account_id->unique_key);
+  register_account_statement.BindInt64(1, account_birthday_block);
+  register_account_statement.BindInt64(2, account_birthday_block);
+  register_account_statement.BindString(3, account_birthday_block_hash);
+
+  if (!register_account_statement.Run()) {
+    transaction.Rollback();
+    NOTREACHED_IN_MIGRATION();
+    return Error{kFailedToExecuteStatement, database_.GetErrorMessage()};
+  }
+
+  if (!transaction.Commit()) {
+    return Error{kFailedToExecuteStatement, database_.GetErrorMessage()};
+  }
+
+  return std::nullopt;
+}
+
+base::expected<ZCashOrchardStorage::AccountMeta, ZCashOrchardStorage::Error>
+ZCashOrchardStorage::GetAccountMeta(mojom::AccountIdPtr account_id) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!EnsureDbInit()) {
+    NOTREACHED_IN_MIGRATION();
+    return base::unexpected(
+        Error{ErrorCode::kDbInitError, database_.GetErrorMessage()});
+  }
+
+  sql::Statement resolve_account_statement(database_.GetCachedStatement(
+      SQL_FROM_HERE,
+      "SELECT account_birthday, latest_scanned_block, "
+      "latest_scanned_block_hash FROM " kAccountMeta " WHERE account_id = ?;"));
+
+  resolve_account_statement.BindString(0, account_id->unique_key);
+
+  if (!resolve_account_statement.Step()) {
+    return base::unexpected(Error{kAccountNotFound, "Account not found"});
+  }
+
+  AccountMeta account_meta;
+  auto account_birthday = ReadUint32(resolve_account_statement, 0);
+  auto latest_scanned_block = ReadUint32(resolve_account_statement, 1);
```

Comment:
Is there any sort of further validation we should be doing on this data here beyond just checking it's a valid uint32? For example, should we check that the latest_scanned_block is sequential in any way or that the account birthday is after a certain valid number?

Reply:
I am not sure we should do this at the data layer. We have some checks in the zcash_shield_sync_service.

Reply:
Ahh, yeah I agree with that. Having looked through to see how the storage and sync service interact again I think this isn't an issue any longer.